File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/application/event/PaymentCancelledEventHandler.java
package com.example.accountservice.payment.application.event;

import static net.logstash.logback.argument.StructuredArguments.kv;

import com.example.accountservice.outbox.OutboxEventHandler;
import com.example.accountservice.outbox.OutboxWriter;
import com.example.accountservice.payment.application.integrationevent.PaymentCancelledIntegrationEventV1;
import com.example.accountservice.payment.domain.PaymentCancelledEvent;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

/**
 * Processes the {@link PaymentCancelledEvent} accumulated in the Outbox, translates it into the
 * {@code payment.cancelled.v1} Integration Event, and stores it. The Account BC subscribes to this
 * and runs a compensating credit (deposit) — a compensating transaction that reverses the amount
 * already deducted.
 */
@Component
@RequiredArgsConstructor
public class PaymentCancelledEventHandler implements OutboxEventHandler {

    private static final Logger log = LoggerFactory.getLogger(PaymentCancelledEventHandler.class);

    private final OutboxWriter outboxWriter;
    private final ObjectMapper objectMapper;

    @Override
    public String eventType() {
        return PaymentCancelledEvent.class.getSimpleName();
    }

    @Override
    public void handle(String payload) throws Exception {
        PaymentCancelledEvent event = objectMapper.readValue(payload, PaymentCancelledEvent.class);
        log.info(
                "Payment cancelled",
                kv("payment_id", event.paymentId()),
                kv("account_id", event.accountId()),
                kv("reason", event.reason()));

        outboxWriter.save(
                PaymentCancelledIntegrationEventV1.EVENT_TYPE,
                new PaymentCancelledIntegrationEventV1(
                        event.paymentId(), event.accountId(), event.amount()));
    }
}

File: implementations/go/examples/internal/application/event/payment_cancelled_event_handler.go
package event

import (
	"context"
	"encoding/json"
	"fmt"

	integrationevent "github.com/example/account-service/internal/application/integration-event"
	"github.com/example/account-service/internal/domain/payment"
)

// PaymentCancelledEventHandler receives the PaymentCancelled domain event
// persisted in the outbox, translates it into a payment.cancelled.v1
// Integration Event, and persists it to the Outbox. The Account BC
// subscribes to this and executes a compensating credit (deposit) — a
// compensating transaction that reverses an amount already debited.
type PaymentCancelledEventHandler struct {
	publisher IntegrationPublisher
}

func NewPaymentCancelledEventHandler(publisher IntegrationPublisher) *PaymentCancelledEventHandler {
	return &PaymentCancelledEventHandler{publisher: publisher}
}

func (h *PaymentCancelledEventHandler) Handle(ctx context.Context, payload []byte) error {
	var evt payment.PaymentCancelled
	if err := json.Unmarshal(payload, &evt); err != nil {
		return fmt.Errorf("unmarshal PaymentCancelled: %w", err)
	}

	ie := integrationevent.PaymentCancelledV1{
		PaymentID:   evt.PaymentID,
		AccountID:   evt.AccountID,
		Amount:      evt.Amount,
		CancelledAt: evt.CancelledAt,
	}
	if err := h.publisher.Publish(ctx, ie.EventName(), ie); err != nil {
		return fmt.Errorf("publish payment.cancelled.v1: %w", err)
	}
	return nil
}
